### ISSUE TITLE
Automatically restart waybar after config change

### ISSUE THREAD

**Issue by Hubro** [Alexays/Waybar#961]
Polybar has this option:

```
       -r, --reload
              Reload the application when the config file has been modified
```

Is there anything equivalent for Waybar?

Currently my workflow is to update the config, run "killall waybar" and then restart it, which is quite time consuming when doing lots of tweaks.

Rather than re-invent the wheel, is there currently a way to achieve this with Waybar? I have been utterly unable to find anything about this after thorough Googling and searching through existing issues, so my feeling is that I am missing something completely obvious.

**Comment by Markaos** [comment]
As a workaround, you could write a script using inotify to restart Waybar whenever the config changes - example (first Google result): https://superuser.com/questions/181517/how-to-execute-a-command-whenever-a-file-changes

**Comment by Hubro** [comment]
> As a workaround, you could write a script using inotify to restart Waybar whenever the config changes - example (first Google result): https://superuser.com/questions/181517/how-to-execute-a-command-whenever-a-file-changes

Yeah my current workaroud looks like this:

`launch-waybar`:

```sh
#!/bin/bash

CONFIG_FILES="$HOME/.config/waybar/config $HOME/.config/waybar/style.css"

trap "killall waybar" EXIT

while true; do
    waybar &
    inotifywait -e create,modify $CONFIG_FILES
    killall waybar
done
```

**Comment by x10an14** [comment]
Thanks @Hubro! I'm currently making use of your solution until something "better" comes along =)

**Comment by hiszd** [comment]
@Hubro how do you execute this from your Sway config?
I am trying to run it using an exec command and it isn't working.
This is at the bottom of my sway config

`bar {
     swaybar_command waybar
}

exec ~/.config/sway/launch_waybar.sh`

**Comment by Hubro** [comment]
@hiszd I have this line at the end of my Sway config:

```
exec --no-startup-id "/bin/bash -c '[[ -f $HOME/autostart.sh ]] && . $HOME/autostart.sh &>> $HOME/autostart.log'"
```

Inside `authstart.sh` I have this line:

```
start launch-waybar
```

And the `start` command is just this script:

```
#!/bin/bash

nohup "$@" &>"/tmp/start-$(basename $1).log" &
```

**Comment by maLi9n** [comment]
I wrote this script  to reload waybar, just change the variables that point to the config and style.css files. I also put in a bind in my hyprland.conf file to run it.

`#!/bin/bash

CONFIG="$HOME/.config/hypr/waybar/config"
STYLE="$HOME/.config/hypr/waybar/style.css"

# This script reloads waybar

# First, we check if waybar is currently running
if [[ $(pgrep -x "waybar") = "waybar" ]];
then
# If waybar is running, we send a killall signal to stop it
    killall waybar
else
# If waybar is not running, we start it
if [[ $(pgrep -x "waybar") = "" ]];
then
    waybar -c $CONFIG -s $STYLE > /dev/null 2>&1 &
fi
fi`

**Comment by Hubro** [comment]
@maLi9n Triple backticks on separate lines for code blocks.

    ```
    This is a code block
    ```

**Comment by HarHarLinks** [comment]
https://github.com/Alexays/Waybar/wiki/FAQ#how-can-I-reload-the-configuration-without-restarting-waybar

**Comment by abhiyandhakal** [comment]
This is my solution. I have this script in exec-once.
```
#!/bin/bash

# start waybar if not started
if ! pgrep -x "waybar" > /dev/null; then
	waybar &
fi

# current checksums
current_checksum_config=$(md5sum ~/.config/waybar/config)
current_checksum_style=$(md5sum ~/.config/waybar/style.css)

# loop forever
while true; do
	# new checksums
	new_checksum_config=$(md5sum ~/.config/waybar/config)
	new_checksum_style=$(md5sum ~/.config/waybar/style.css)

	# if checksums are different
	if [ "$current_checksum_config" != "$new_checksum_config" ] || [ "$current_checksum_style" != "$new_checksum_style" ]; then
		# kill waybar
		killall waybar

		# start waybar
		waybar &

		# update checksums
		current_checksum_config=$new_checksum_config
		current_checksum_style=$new_checksum_style
	fi
done
```

This kills and restarts waybar every time the config changes. The change in config is detected by matching the checksums.

**Comment by stratosgear** [comment]
So your loop constantly calculates checksums forever...?  Seems highly inefficient, borderline abusive to the system...

I would expect that waybar would implement something along the lines of a [SIGHUP](https://en.wikipedia.org/wiki/Signal_(IPC)#SIGHUP) that seems to be the standard for these kind of operations...

The rest of the scripts offered above, that kill/restart waybar, have an annoying refresh of the screen (at least on hyprland when the desktop is redrawn to occupy the waybar space on top).

My usecase is to change the colors of the styles (changed with pywal)  something that should not cause any geometry changes so  the switch should be almost "transparent".

**Comment by ssamin69** [comment]
> This is my solution. I have this script in exec-once.
> 
> ```
> #!/bin/bash
> 
> # start waybar if not started
> if ! pgrep -x "waybar" > /dev/null; then
> 	waybar &
> fi
> 
> # current checksums
> current_checksum_config=$(md5sum ~/.config/waybar/config)
> current_checksum_style=$(md5sum ~/.config/waybar/style.css)
> 
> # loop forever
> while true; do
> 	# new checksums
> 	new_checksum_config=$(md5sum ~/.config/waybar/config)
> 	new_checksum_style=$(md5sum ~/.config/waybar/style.css)
> 
> 	# if checksums are different
> 	if [ "$current_checksum_config" != "$new_checksum_config" ] || [ "$current_checksum_style" != "$new_checksum_style" ]; then
> 		# kill waybar
> 		killall waybar
> 
> 		# start waybar
> 		waybar &
> 
> 		# update checksums
> 		current_checksum_config=$new_checksum_config
> 		current_checksum_style=$new_checksum_style
> 	fi
> done
> ```
> 
> This kills and restarts waybar every time the config changes. The change in config is detected by matching the checksums.

just use this bro

`killall -SIGUSR2 waybar`
make a script to run this with pywal but seems awkward to make a script for few words so maybe 
put this in along with other scripts if you have for pywal .

**Comment by abhiyandhakal** [comment]
For the above comments, I still want hot restart, so that I don't have to restart the waybar myself (even with the keybinds...). So, it is still necessary for my use-case.

**Comment by stratosgear** [comment]
You do not need a separate keybind (although, you "might" have one, it doesn't hurt).

You simply use a `killall -SIGUSR2 waybar` at the end of any script that you might have that automatically changes things in the background (like rotating wallpapers, updating colors etc).  You will **not** be manually restarting waybar, the script will.  In case you manually make changes to the waybar files (editing styles) you could indeed use the keybind to reload.

In any case, what works for me does not imply that works for everyone else, obviously.  I just don't see the need for custom scripts when waybar already makes allowances for cases like this!

Enjoy!

**Comment by abhiyandhakal** [comment]
Yeah that would be really effective since it doesn't require another infinite loop... I don't remember if I have any such scripts... I will surely do that if I remember any. Thanks!

**Comment by dinhokusanagi** [comment]
> Esta é a minha solução. Eu tenho esse script em exec-once.
> 
> ```
> #!/bin/bash
> 
> # start waybar if not started
> if ! pgrep -x "waybar" > /dev/null; then
> 	waybar &
> fi
> 
> # current checksums
> current_checksum_config=$(md5sum ~/.config/waybar/config)
> current_checksum_style=$(md5sum ~/.config/waybar/style.css)
> 
> # loop forever
> while true; do
> 	# new checksums
> 	new_checksum_config=$(md5sum ~/.config/waybar/config)
> 	new_checksum_style=$(md5sum ~/.config/waybar/style.css)
> 
> 	# if checksums are different
> 	if [ "$current_checksum_config" != "$new_checksum_config" ] || [ "$current_checksum_style" != "$new_checksum_style" ]; then
> 		# kill waybar
> 		killall waybar
> 
> 		# start waybar
> 		waybar &
> 
> 		# update checksums
> 		current_checksum_config=$new_checksum_config
> 		current_checksum_style=$new_checksum_style
> 	fi
> done
> ```
> 
> Isso mata e reinicia o waybar toda vez que a configuração é alterada. A mudança na configuração é detectada pela correspondência das somas de verificação.

nice!
worked perfectly on hyperland, thank you

**Comment by omnigenous** [comment]
@Hubro Can you please explain what does line with `trap` does? I'm trying to modify your script to use `killall -SIGUSR2 waybar`

**Comment by Hubro** [comment]
@omnigenous It runs "killall waybar" when the script exits, so waybar doesn't stay open in the background. I don't remember if it's necessary, but it works.

**Comment by annahri** [comment]
Doing `killall -SIGUSR2 waybar` does "refresh" the bar but does not reflect the changes in the config.

Edit: My comment is no longer relevant. USR2 signal now reloads everything.

**Comment by jasper-at-windswept** [comment]
Hey I was wondering does anyone know if there is anyway i can just refresh the CSS and not the whole bar as that results in it flashing as it re-opens each time.

For context I am trying to add a element that displays the cover art of the currently playing song on MPD.
I have come up with this script which just runs at the startup of my PC and symlinks the cover art to a local file where Waybar can access it.

Before anyone says yes i have set `"reload_style_on_change": true` and that updates the size of the element but strangely not the actual image, it justs remains the same as whatever the image was when waybar started.

Script:
```sh
playerctl --follow metadata | while read -r metadata; do
    image_url=$(echo $metadata | grep "mpris:artUrl" | sed -e 's/.*file:\/\///g')
    if [ -n "$image_url" ]; then
        if [[ "$image_url" == *".png" ]]; then
            ln -fs "$image_url" ~/.config/waybar/cover.png
            rm ~/.config/waybar/cover.jpg
        else
            ln -fs "$image_url" ~/.config/waybar/cover.jpg
        fi
        pkill -SIGUSR2 waybar
        echo "Symlink created for $image_url"
    fi
done
```

And below here is the CSS that makes it work.
```css
  #mpd {
    padding: 5 55px;
    border-radius: 12px;
    background-image: url("cover.jpg"), url("cover.png");
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;
    color: #FFF;
    margin-top: 8px;
    margin-bottom: 8px;
    margin-right: 2px;
    margin-left: 2px;
    transition: all 0.3s ease;
    font-weight: 900;
  }
```

Help greatly appreciated.

**Comment by tfemby** [comment]
If you just want to use `killall -SIGUSR2 waybar` (where it just works on css updates), you could use a systemd-path. I have this in my `~/.config/systemd/user/`

reload-waybar.service
```
[Unit]
Description=Reload waybar css

[Service]
Type=oneshot
ExecStart=killall -SIGUSR2 waybar
ExecStart=notify-send "Reloaded waybar"
```

reload-waybar.path
```
[Unit]
Description=Watch waybar for changes

[Path]
PathModified=%h/.config/waybar/style.css
```

Then start it with `systemctl --user start reload-waybar.path` which you could easily put into your wm's autostart or just use systemd to enable it.

**Comment by htayj** [comment]
for posterity, here is simple script using the USR2 signal. it launches waybar and hotreloads when any of the config files change and kills it if you exit the script:

```bash
#!/bin/bash
waybar&
trap "killall waybar" EXIT
while inotifywait -r -e create,modify .config/waybar/*; do killall -SIGUSR2 waybar; done
```

**Comment by enoryw** [comment]
USR2 signal reload waybar config for me, not just css. At least for changing position of a bar or commenting out a module.

For those who claim it only works for CSS changes, what doesn't work update in the config when sending the signal? It's as good of a hot restart.

**Comment by marcelmanz** [comment]
For me the best solution is to use this:
```
exec = pkill waybar; waybar &
```
in **.config/hypr/hyprland.conf**

then every time that hyprland is reloaded it will reload also the waybar.

**Comment by dividebysandwich** [comment]
Agree with @enoryw , USR2 signal is all that's needed for both config and CSS changes.

**Comment by bitfuzz** [comment]
Hello, I am new to open source contributing so please be easy one me.
Id like to work on implementing native auto-reload functionality for Waybar.

I see there are several workarounds posted here , but having this built-in would be valuable for the same reasons Polybar includes the `-r` flag.

Before I start working on a PR, I wanted to check:

1. Would a PR implementing this feature be accepted?
2. Are there any specific implementation preferences? I'm considering:
   - Using inotify to watch config files (similar to existing workarounds)
   - Adding a `-r/--reload` CLI flag to match Polybar's interface
   - Optionally making it configurable via the config file itself

3. Should this reload both config.json AND style.css, or would you prefer separate handling?

Looking forward to any guidance before I dive into the codebase!

**Comment by Alexays** [comment]
Yes — Waybar reloads in place on `SIGUSR2`, no need to kill/restart:
```sh
killall -SIGUSR2 waybar
```
That does a full reload of both `config` and `style.css` without relaunching the process (SIGUSR2 → RELOAD; for reference SIGUSR1 toggles visibility). Just hook it into whatever changes your files — a pywal script, an editor hook, or a tiny watcher:
```sh
while inotifywait -e modify ~/.config/waybar/config ~/.config/waybar/style.css; do
  killall -SIGUSR2 waybar
done
```
This is also covered in the FAQ. Closing, but happy to reopen if needed. 🙏
